drivers/ethernet/phy/phy_microchip_t1s.c

```diff
@@ -0,0 +1,561 @@
+/*
+ * Copyright (c) 2024 Microchip Technology Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT microchip_t1s_phy
+
+#include <zephyr/kernel.h>
+#include <zephyr/net/phy.h>
+#include <zephyr/net/mii.h>
+#include <zephyr/net/mdio.h>
+#include <zephyr/drivers/mdio.h>
+
+#define LOG_MODULE_NAME phy_mc_t1s
+#define LOG_LEVEL       CONFIG_PHY_LOG_LEVEL
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(LOG_MODULE_NAME);
+
+/* Support: Microchip Phys:
+ * lan8650/1 Rev.B0/B1 Internal PHYs
+ */
+/* Both Rev.B0 and B1 clause 22 PHYID's are same due to B1 chip limitation */
+#define PHY_ID_LAN865X_REVB 0x0007C1B3
+
+/* Configuration param registers */
+#define LAN865X_REG_CFGPARAM_ADDR    0x00D8
+#define LAN865X_REG_CFGPARAM_DATA    0x00D9
+#define LAN865X_REG_CFGPARAM_CTRL    0x00DA
+#define LAN865X_REG_STS2             0x0019
+#define LAN865X_CFGPARAM_READ_ENABLE BIT(1)
+
+/* Collision detection enable/disable registers */
+#define LAN86XX_DISABLE_COL_DET   0x0000
+#define LAN86XX_ENABLE_COL_DET    0x8000
+#define LAN86XX_COL_DET_MASK      0x8000
+#define LAN86XX_REG_COL_DET_CTRL0 0x0087
+
+/* LAN865x Rev.B0/B1 configuration parameters from AN1760
+ * As per the Configuration Application Note AN1760 published in the below link,
+ * https://www.microchip.com/en-us/application-notes/an1760
+ * Revision F (DS60001760G - June 2024)
+ */
+static const uint32_t lan865x_revb_fixup_registers[17] = {
+	0x00D0, 0x00E0, 0x00E9, 0x00F5, 0x00F4, 0x00F8, 0x00F9, 0x0081, 0x0091,
+	0x0043, 0x0044, 0x0045, 0x0053, 0x0054, 0x0055, 0x0040, 0x0050,
+};
+
+static const uint16_t lan865x_revb_fixup_values[17] = {
+	0x3F31, 0xC000, 0x9E50, 0x1CF8, 0xC020, 0xB900, 0x4E53, 0x0080, 0x9660,
+	0x00FF, 0xFFFF, 0x0000, 0x00FF, 0xFFFF, 0x0000, 0x0002, 0x0002,
+};
+
+static const uint16_t lan865x_revb_fixup_cfg_regs[2] = {
+	0x0084,
+	0x008A,
+};
+
+static const uint32_t lan865x_revb_sqi_fixup_regs[12] = {
+	0x00B0, 0x00B1, 0x00B2, 0x00B3, 0x00B4, 0x00B5,
+	0x00B6, 0x00B7, 0x00B8, 0x00B9, 0x00BA, 0x00BB,
+};
+
+static const uint16_t lan865x_revb_sqi_fixup_values[12] = {
+	0x0103, 0x0910, 0x1D26, 0x002A, 0x0103, 0x070D,
+	0x1720, 0x0027, 0x0509, 0x0E13, 0x1C25, 0x002B,
+};
+
```

Comment:
How the above changes are corresponding to:
https://github.com/zephyrproject-rtos/zephyr/pull/80639/commits/f0f91a8c8686c7cf761ca08e51e4b34d0a0f7a88

Reply:
Ok, but we can only use address and value parameters because as PHY driver point of view it is only MMD of 0x1F and MAC driver will do the job of mapping the corresponding MMD to MMS as MMS is handled only in the MAC driver. Also I used the same approach used in the Linux driver.

https://elixir.bootlin.com/linux/v6.13-rc1/source/drivers/net/phy/microchip_t1s.c

Still do you want me to follow the above approach OR shall we keep as it is for now?

Reply:
My point here is that when you look on:
https://ww1.microchip.com/downloads/aemDocuments/documents/AIS/ApplicationNotes/ApplicationNotes/LAN8650-1-Configuration-Appnote-60001760.pdf
the `Write Configuration into Registers` chapter `Table1` - you have a sequence to be written. It is easier to maintenance (as those values have changed also recently in this document) and implement than by using the approach from Linux driver.
The above mentioned link to [Linux driver](https://elixir.bootlin.com/linux/v6.12.1/source/drivers/net/phy/microchip_t1s.c#L63) starts the configuration from the middle of the `Table1` (and seems to be out of sync with newest AN1760).

I would like to have the code as close to Microchip's recommendations as possible.

Reply:
> My point here is that when you look on: https://ww1.microchip.com/downloads/aemDocuments/documents/AIS/ApplicationNotes/ApplicationNotes/LAN8650-1-Configuration-Appnote-60001760.pdf the `Write Configuration into Registers` chapter `Table1` - you have a sequence to be written. It is easier to maintenance (as those values have changed also recently in this document) and implement than by using the approach from Linux driver. The above mentioned link to [Linux driver](https://elixir.bootlin.com/linux/v6.12.1/source/drivers/net/phy/microchip_t1s.c#L63) starts the configuration from the middle of the `Table1` (and seems to be out of sync with newest AN1760).
I think you are looking at the old driver which was supporting for the previous version of AN1760. Please refer the latest one mainlined recently below which follows exactly the same sequence mentioned in the latest AN1760.
[Latest Linux driver v6.13](https://elixir.bootlin.com/linux/v6.13-rc1/source/drivers/net/phy/microchip_t1s.c#L76)

Hope this clarifies?
> 
> I would like to have the code as close to Microchip's recommendations as possible.

Reply:
From the aforementioned document:
`Table 1 lists a series of parameters that must be configured with the given values and in the order listed. `
The approach took in the Linux driver is not doing the configuration in recommended "order" - for example:
`W 0x4 0x0084 cfgparam1` is done last.

The code, which is already in Zephyr (https://github.com/zephyrproject-rtos/zephyr/commit/f0f91a8c8686c7cf761ca08e51e4b34d0a0f7a88) - follows exactly the recommendations.

Reply:
Ah ok, you mean the order. As per the internal discussion we had regarding this order, implemented in Linux as it might not break anything. I will double check with design team again. Meanwhile as per your suggestion, let's stick with the order provided in the AN1760. I will update it.